### PR TITLE
[MNT] Remove upper bound on python_requires

### LINKED ISSUE #780: Remove upper bound on python_requires
:wave: At the moment the project has an upper bound on `python_requires`

https://github.com/glotzerlab/signac/blob/aed17f324b3e1f5dfb3328cff6c2ab73ca434929/setup.py#L63

While `signac` is a framework and not a "pure" library, I would still consider it to [not be an application](https://iscinumpy.dev/post/app-vs-library/). I would recommend removing this upper bound on Python 4 for the reasons summarized in Henry's [detailed blog post on the subject](https://iscinumpy.dev/post/bound-version-constraints/) and because the Python Steering Council has made it clear that if there ever is a Python 4 it will not be as radical of a break as Python 2 -> Python 3.

### PR DESCRIPTION
**Pull request by matthewfeickert**
## Description

* Remove upper bound of Python 4 on `python_requires` as advocated for in Henry Schreiner's blog post "[Should You Use Upper Bound Version Constraints?](https://iscinumpy.dev/post/bound-version-constraints/)". 

## Motivation and Context
This change follows best practices for libraries and frameworks as generally recommended by the PyPA.

Resolves #780

## Checklist:

- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] The [package documentation](https://github.com/glotzerlab/signac/tree/master/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added any related issue and pull request numbers for future reference.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
Note: the same thing should also be done for additional glotzerlab repositories.

signac-flow, signac-dashboard, rowan, gsd (uses `~=3.6` which would not match 4), garnett, hoomdpy-addons.

I can help with signac packages but leave the rest to other maintainers. cc: @b-butler @joaander @vyasr

File: changelog.txt
Comment (matthewfeickert) [reply changelog.txt:52]: I don't personally care, but I included it as given the rationale of a PyPA member who influences this for all Python users seemed relevant.

Comment (tommy-waltmann) [reply changelog.txt:52]: I don't have a strong opinion either, but I thought I'd mention it. I'll defer to the signac maintainers on this.

Comment (vyasr) [reply changelog.txt:52]: I don't think it needs to be in the changelog. It'll be in the commit log (the PR description gets put in the squash commit's message) if anyone is sufficiently interested, so I'd prefer to keep the changelog lean.

Comment (matthewfeickert) [reply changelog.txt]: SGTM. I've fixed it up and force pushed it back up. Thanks for raising it and reviewing @tommy-waltmann and @vyasr!

**Review by tommy-waltmann — APPROVED** [review]

**Pull request merged**
